### PR TITLE
Add machine type availability checks to slurm-gcp-v6-nodeset

### PR DESCRIPTION
**Pull request by annuay-google**
### Issue
Blueprint yaml files allow users to specify machine and zone combinations that do not exist. Infra is provisioned by Terraform during `./ghpc deploy` , but autoscaling may fail later if capacity is not found by bulk insert APIs in zones specified

### Approach
Created a Terraform precondition to verify that the machine type is available in at least one zone. In case they're not, there is no feasible way for bulk insert to allocate machines, and terraform will exit during plan/apply

### Testing
Ran `./ghpc create`, `./ghpc deploy` on this configuration. The specified machine type does not exist in both zones `us-central1-b` and `us-central1-c`. Verified that Terraform exits and error message is as expected

config:
```yaml
vars:
  deployment_name: hpc-slurm
  region: us-central1
  zone: us-central1-b

deployment_groups:
- group: primary
  modules:
  - id: h3_nodeset
    source: community/modules/compute/schedmd-slurm-gcp-v6-nodeset
    use: [network]
    settings:
      node_count_dynamic_max: 20
      # Note that H3 is available in only specific zones. https://cloud.google.com/compute/docs/regions-zones
      machine_type: h3-standard-88
      # H3 does not support pd-ssd and pd-standard
      # https://cloud.google.com/compute/docs/compute-optimized-machines#h3_disks
      disk_type: pd-balanced
      bandwidth_tier: gvnic_enabled
      allow_automatic_updates: false
      zones:
      - us-central1-c
```
output:
```
Testing if deployment group hpc-slurm/primary requires adding or changing cloud infrastructure
Error: exit status 1

Error: Resource precondition failed

  on modules/embedded/community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf line 193, in resource "terraform_data" "machine_type_zone_validation":
 193:       condition     = length(local.zones_with_machine_type) > 0
    ├────────────────
    │ local.zones_with_machine_type is empty tuple

machine type h3-standard-88 is not available in any of the zones
["us-central1-b","us-central1-c"]. To list zones in which it is available, run:

gcloud compute machine-types list --filter="name=h3-standard-88"

Hint: terraform plan for deployment group hpc-slurm/primary failed
```
Added 1 valid zone (`us-central1-a`) and verified that infra is correctly provisioned

### Submission Checklist

Please take the following actions before submitting this pull request.

- [x] Fork your PR branch from the Toolkit "develop" branch (not main)
- [x] Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
- [x] Confirm that "make tests" passes all tests
- [ ] Add or modify unit tests to cover code changes
- [x] Ensure that unit test coverage remains above 80%
- [x] Update all applicable documentation
- [x] Follow Cluster Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf
Comment (mr0re1) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: Check is available starting Terraform 1.5.0 . AFAIK we are supporting terraform 1.2.0.

See `Makefile`: `MIN_TERRAFORM_VERSION=1.2 # for deploying modules`

Comment (annuay-google) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: Yes, checks are available with only TF 1.5+. I had discussed a similar concern regarding TF version with Tom (that turned out to be a non-issue). He had clarified then that we should be able to use Terraform 1.5

@tpdownes would it be okay for us to bump up the minimum version to 1.5?

Comment (mr0re1) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: LGTM, feel free to land if you get a green light in regard of terraform version.

**Review by mr0re1 — DISMISSED** [review]

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf
Comment (annuay-google) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: ack, will wait for Tom

Comment (nick-stroud) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: I agree that TF 1.5 is acceptable, but we should wait to put this in until https://github.com/GoogleCloudPlatform/cluster-toolkit/pull/2772 has been rolled back.

Comment (tpdownes) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: Completed in #2961

Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: 1. You have the [wrong syntax for checks](https://developer.hashicorp.com/terraform/language/checks#syntax). The nice thing about checks is that data sources are "scoped" inside the check block which allows them to fail.

2. I think you are using the wrong condition. The assertion we want is "at least 1 zone has the machine type" not "there are 0 zones without the machine type". Both your examples should succeed.

3. Despite (1), it's interesting to me that the condition is failing in the check block rather than the data source. I would think that it would fail at the data source stage. If the google_compute_machine_types data source can successfully return the empty list, it would actually be more desirable to **fail** as a precondition than to warn with a check.

Can you please verify the behavior of the data source on a to confirm that it can return empty without failing?

e.g.

```hcl
data "google_compute_machine_types" "machine_types_by_zone" {
  for_each = toset(["us-central1-b", "us-central1-c"])
  filter   = "name = \"h3-standard-88\""
  zone     = each.value
}
```

Run `terraform state list` and `terraform state show data.google_compute_machine_types.machine_types_by_zone[\"us-central1-b\"]`. If this returns the empty list after `terraform apply`, then we are safe to fail on the data source instead of warning in a check block. So long as you fix the boolean logic (2).

**Review by tpdownes — CHANGES_REQUESTED** [review]

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf
Comment (annuay-google) [reply community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:190]: Verified that 3 works. The data source does not fail.

Updated the code to use `precondition` with `terraform_data` resource as discussed over the chat. If none of the zones have the machine type available, Terraform will throw an error and exit.

Tested the code on success and failure scenarios (updated the PR description).

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:191]: You should be chaining this resource into the graph dependencies.

https://developer.hashicorp.com/terraform/language/resources/terraform-data

I suggest adding:

```hcl
  input = var.machine_type
```

and then replace line 80 with

```
    machine_type = terraform_data.machine_type_zone_validation.output
```

**Review by tpdownes — CHANGES_REQUESTED** [review]
To make the terraform_data resource and precondition useful, it should be inserted into the terraform resource graph.

**Review by tpdownes — CHANGES_REQUESTED** [review]
Please also change versions.tf to require terraform 1.4

**Comment by annuay-google** [comment]
> Please also change versions.tf to require terraform 1.4

Done

File: community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf
Comment (tpdownes) [inline community/modules/compute/schedmd-slurm-gcp-v6-nodeset/main.tf:195]: A good suggestion was made to me to help the user to the right answer. I suggest the following:


 to guide the user to select
```suggestion
      error_message = <<-EOT
        machine type ${var.machine_type} is not available in any of the zones ${jsonencode(local.zones)}". To list zones in which it is available, run:

        gcloud compute machine-types list --filter="name=${var.machine_type}"
        EOT
```

**Review by tpdownes — CHANGES_REQUESTED** [review]
Please add one final helpful hint to the user. Then squash your commits and this should be ready!

**Comment by annuay-google** [comment]
> Please add one final helpful hint to the user. Then squash your commits and this should be ready!

Done. Verified the error as well
```
machine type h3-standard-88 is not available in any of the zones
["us-central1-b"]". To list zones in which it is available, run:

gcloud compute machine-types list --filter="name=h3-standard-88"
```

**Review by tpdownes — APPROVED** [review]
This is a big improvement in user experience. Thank you for the submission!

**Pull request merged**
